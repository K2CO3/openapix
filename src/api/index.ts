export { OpenApiProps, OpenApiBaseProps, Validator } from './api-props';
export { OpenApi } from './api';
