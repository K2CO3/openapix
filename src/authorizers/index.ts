export { AuthorizerConfig, AuthorizerExtensions } from './authorizer';
export { LambdaAuthorizer, LambdaAuthorizerProps } from './lambda';
export { CognitoUserPoolsAuthorizer, CognitoUserPoolsAuthorizerProps } from './cognito';
