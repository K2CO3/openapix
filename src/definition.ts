import * as fs from 'fs';
import * as cdk from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as assets from 'aws-cdk-lib/aws-s3-assets';
import { Construct } from 'constructs';
import { omit, set, get } from 'lodash';import { OpenApiXIntegration } from './integrations/base';
import { OpenApiXSource } from './source';
import { resolveLambdaIntegrationUri } from './x-amazon-integration/lambda';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const omitDeep = require('omit-deep-lodash');


type Method = 'ANY'|'DELETE'|'GET'|'HEAD'|'OPTIONS'|'PATCH'|'POST'|'PUT';
type OpenApiMethodIntegrations = Record<string, OpenApiXIntegration> // TODO add validation for method
type OpenApiPathIntegrations = Record<string, OpenApiMethodIntegrations>


export interface OpenApiXDefinitionProps {
  readonly upload?: boolean;
  readonly source: string | OpenApiXSource;
  readonly integrations?: OpenApiPathIntegrations;
  readonly injectPaths?: Record<string, any>;
  readonly rejectPaths?: string[];
  readonly rejectDeepPaths?: string[];
  readonly customAuthorizer?: lambda.IFunction;
}

export class OpenApiXDefinition extends apigateway.ApiDefinition {
  private definition?: any;
  private s3Location?: apigateway.ApiDefinitionS3Location;
  private upload: boolean;
  private scope: Construct;


  constructor(scope: Construct, props: OpenApiXDefinitionProps) {
    super();

    this.scope = scope;

    const { upload = false, source, integrations = {}, injectPaths = {}, rejectPaths = [], rejectDeepPaths = [], customAuthorizer } = props;

    this.upload = upload;

    let schemaSource: OpenApiXSource;

    if (typeof source === 'string') {
      schemaSource = OpenApiXSource.fromAsset(source);
    } else {
      schemaSource = source;
    }

    // TODO: this all should be reworked, too much mutating and confusing stuff
    const sourceDefinition = schemaSource;

    // hard coded custom authorizer, should be refactored
    if (customAuthorizer) {
      const authorizerUri = resolveLambdaIntegrationUri(scope, customAuthorizer);

      set(sourceDefinition.definition, 'components.securitySchemes.customAuthorizer', {
        'type': 'apiKey', // Required and the value must be "apiKey" for an API Gateway API.
        'name': 'Authorization', // The name of the header containing the authorization token.
        'in': 'header', // Required and the value must be "header" for an API Gateway API.
        'x-amazon-apigateway-authorizer': {
          identitySource: 'method.request.header.Authorization',
          authorizerUri,
          authorizerResultTtlInSeconds: 300,
          type: 'request',
          enableSimpleResponses: false,
        },
        'x-amazon-apigateway-authtype': 'Custom authorizer',
      });
    }

    Object.keys(integrations).map(path => {
      this.ensurePath(sourceDefinition.definition, path);
      const integrationsForPath = integrations[path];
      Object.keys(integrationsForPath).map(m => {
        const method = m.toLowerCase();
        this.ensureMethod(sourceDefinition.definition, path, method);
        this.ensureNoIntegration(sourceDefinition.definition, path, method);

        const integration = integrationsForPath[method.toUpperCase() as Method]!;

        //schemaJson.paths![path][method]['x-amazon-apigateway-integration'] = integration.xAmazonIntegration;
        set(sourceDefinition.definition, `paths['${path}']['${method}']['x-amazon-apigateway-integration']`, integration.xAmazonIntegration);

        if (customAuthorizer) {
          set(sourceDefinition.definition, `paths['${path}']['${method}']['security']`, [{ customAuthorizer: [] }]);
        }
      });
    });

    //Object.keys(injectPaths).forEach(path => set(schemaJson, path, injectPaths[path]));
    sourceDefinition.definition = this.injectPaths(sourceDefinition.definition, injectPaths);

    //rejectPaths.forEach(path => omit(schemaJson, path));
    sourceDefinition.definition = this.rejectPaths(sourceDefinition.definition, rejectPaths);

    //rejectDeepPaths.forEach(path => omitDeep(schemaJson, path));
    sourceDefinition.definition = this.rejectDeepPaths(sourceDefinition.definition, rejectDeepPaths);

    const newSchema = sourceDefinition.definition;

    if (this.upload) {
      const newSchemaPath = __dirname + 'open-api-schema.compiled.yaml';
      fs.writeFileSync(newSchemaPath, JSON.stringify(newSchema), 'utf-8');
      const asset = new assets.Asset(scope, 'DefinitionAsset', {
        path: newSchemaPath,
      });

      this.s3Location = {
        bucket: asset.bucket.bucketArn,
        key: asset.s3ObjectKey,
      };
    } else {
      this.definition = newSchema;
    }
  }

  bind(_scope: Construct): apigateway.ApiDefinitionConfig {

    if (this.upload) {
      return {
        s3Location: this.s3Location,
      };
    }

    return {
      inlineDefinition: this.definition,
    };
  }

  private ensurePath(schemaJson: any, path: string): void {
    const value = get(schemaJson, `paths[${path}]`);

    if (typeof value === 'undefined') {
      cdk.Annotations.of(this.scope).addError(`OpenAPI schema is missing path for: ${path}`);
    }
  }

  private ensureMethod(schemaJson: any, path: string, method: string): void {
    const value = get(schemaJson, `paths[${path}][${method}]`);

    if (typeof value === 'undefined') {
      cdk.Annotations.of(this.scope).addError(`OpenAPI schema is missing method ${method} for path: ${path}`);
    }
  }

  private ensureNoIntegration(schemaJson: any, path: string, method: string): void {
    const value = get(schemaJson, `paths[${path}][${method}]['x-amazon-apigateway-integration']`);
    if (typeof value !== 'undefined') {
      cdk.Annotations.of(this.scope).addError(`OpenAPI schema already has x-amazon-apigateway-integration configuration for method ${method} in path: ${path}`);
    }
  }

  private injectPaths(schemaJson: any, injectPaths: Record<string, any> = {}) {
    Object.keys(injectPaths).forEach(path => set(schemaJson, path, injectPaths[path]));
    return schemaJson;
  }

  private rejectPaths(schemaJson: any, rejectPaths: string[] = []) {
    return omit(schemaJson, rejectPaths);
  }

  private rejectDeepPaths(schemaJson: any, rejectDeepPaths: string[] = []) {
    return omitDeep(schemaJson, ...rejectDeepPaths);
  }
}
