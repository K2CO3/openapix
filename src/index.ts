export { OpenApi, OpenApiProps } from './api';
export { OpenApiDefinition, OpenApiDefinitionProps, Validator, SecuritySchemes, XAmazonApigatewayRequestValidator, Authorizer } from './definition';
//export { Integration } from './integration/base';
export { LambdaIntegration, LambdaIntegrationOptions } from './integration/lambda';
export { AwsIntegration, AwsIntegrationProps } from './integration/aws';
export { HttpIntegration, HttpIntegrationProps } from './integration/http';
export { MockIntegration, MockIntegrationProps } from './integration/mock';
export { CorsIntegration, CorsIntegrationProps } from './integration/cors';
export { Schema } from './schema';

export { XAmazonApigatewayAuthorizer } from './x-amazon-apigateway/authorizer';
export { XAmazonApigatewayAuthType } from './x-amazon-apigateway/authtype';
