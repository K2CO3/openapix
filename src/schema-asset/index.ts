import { writeFileSync } from 'fs';
import { tmpdir } from 'os';
import { join } from 'path';
import { Asset } from 'aws-cdk-lib/aws-s3-assets';
import { paramCase } from 'change-case';
import { Construct } from 'constructs';


export interface SchemaAssetProps {
  dir?: string;
}

export class SchemaAsset extends Asset {

  private static resolveDir(props?: Pick<SchemaAssetProps, 'dir'>): string {
    return props?.dir || join(tmpdir(), '@alma-cdk', 'openapix');
  }

  private static resolveFileName(scope: Construct, id: string): string {
    return `${paramCase(scope.node.path)}-${paramCase(id)}.schema.yaml`;
  }

  private static resolveFilePath(dirPath: string, fileName: string): string {
    return join(dirPath, fileName);
  }

  private static write(filePath: string, definition: any): void {
    writeFileSync(filePath, JSON.stringify(definition), 'utf-8');
  }

  constructor(scope: Construct, id: string, definition: any, props?: SchemaAssetProps) {

    const dirPath = SchemaAsset.resolveDir(props);
    const fileName = SchemaAsset.resolveFileName(scope, id);
    const filePath = SchemaAsset.resolveFilePath(dirPath, fileName);
    SchemaAsset.write(filePath, definition);

    super(scope, id, {
      path: filePath,
    });
  }
}
