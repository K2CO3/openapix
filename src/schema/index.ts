export { Schema } from './schema';
export * from './props';
export * from './idocument';
