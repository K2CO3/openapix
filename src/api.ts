import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as iam from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';
import { omit } from 'lodash';
import { OpenApiXDefinition, OpenApiXDefinitionProps } from './definition';
import { OpenApiXIntegration } from './integrations/base';
import { OpenApiXLambda } from './integrations/lambda';

export interface OpenApiXProps extends OpenApiXDefinitionProps {
  readonly restApiProps: apigateway.RestApiBaseProps;
}

export class OpenApiX extends Construct {
  public readonly api: apigateway.IRestApi;

  constructor(scope: Construct, id: string, props: OpenApiXProps) {
    super(scope, id);

    const { restApiProps } = props;

    const apiDefinition = new OpenApiXDefinition(this, omit(props, 'restApiProps'));

    const api = new apigateway.SpecRestApi(this, 'SpecRestApi', {
      apiDefinition,
      ...restApiProps,
    });

    this.api = api;

    this.grantLambdaInvokes(props.integrations);
  }

  grantLambdaInvokes(pathIntegrations: OpenApiXDefinitionProps['integrations']) {
    if (!pathIntegrations) return;

    const apiGatewayInstance = new iam.ServicePrincipal(
      'apigateway.amazonaws.com',
      {
        conditions: [
          {
            ArnLike: {
              'AWS:SourceArn': `${this.api.arnForExecuteApi}/*`,
            },
          },
        ],
      },
    );

    Object.keys(pathIntegrations).forEach(path => {
      const methodIntegrations = pathIntegrations[path];
      Object.keys(methodIntegrations).forEach(method => {
        const methodIntegration = methodIntegrations[method];
        const isLambdaIntegration = (integration: OpenApiXLambda | OpenApiXIntegration):integration is OpenApiXLambda => {
          return !!(integration as OpenApiXLambda).fn;
        };

        if (isLambdaIntegration(methodIntegration)) {
          methodIntegration.fn.grantInvoke(apiGatewayInstance);
        }
      });
    });
  }
}
