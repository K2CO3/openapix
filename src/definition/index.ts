import * as fs from 'fs';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as assets from 'aws-cdk-lib/aws-s3-assets';
import { Construct } from 'constructs';
import { Integration } from '../integrations/base';
import { Schema } from '../schema';
import { SecuritySchemes } from '../security-schemes/interfaces/security-schemes';
import { addError } from '../errors/add';
import { Authorizer } from '../security-schemes/interfaces/authorizer';
import { XAmazonApigatewayRequestValidator } from '../x-amazon-apigateway/request-validators';
import { CorsIntegration } from '../integrations/cors';
// eslint-disable-next-line @typescript-eslint/no-require-imports
const omitDeep = require('omit-deep-lodash');

type Method = 'ANY'|'DELETE'|'GET'|'HEAD'|'OPTIONS'|'PATCH'|'POST'|'PUT';
type OpenApiMethodIntegrations = Record<string, Integration> // TODO add validation for method
type OpenApiPathIntegrations = Record<string, OpenApiMethodIntegrations>

export { SecuritySchemes } from '../security-schemes/interfaces/security-schemes';
export { Authorizer } from '../security-schemes/interfaces/authorizer';
export { XAmazonApigatewayRequestValidator } from '../x-amazon-apigateway/request-validators';

export interface OpenApiDefinitionProps {
  readonly upload?: boolean;
  readonly source: string | Schema;
  readonly integrations?: OpenApiPathIntegrations;
  readonly injections?: Record<string, any>;
  readonly rejections?: string[];
  readonly rejectionsDeep?: string[];
  readonly authorizers?: SecuritySchemes;
  readonly validators?: Record<string, Validator>;
  readonly cors?: CorsIntegration;
}

export interface Validator extends XAmazonApigatewayRequestValidator {
  readonly default?: boolean;
}

export class OpenApiDefinition extends apigateway.ApiDefinition {
  private definition?: any;
  private s3Location?: apigateway.ApiDefinitionS3Location;
  private readonly upload: boolean;
  private readonly scope: Construct;
  private readonly source: Schema;
  private readonly cors?: CorsIntegration;


  constructor(scope: Construct, props: OpenApiDefinitionProps) {
    super();
    const {
      upload = false,
      source,
      authorizers = {},
      integrations = {},
      validators = {},
      injections = {},
      rejections = [],
      rejectionsDeep = [],
      cors,
    } = props;

    this.scope = scope;
    this.upload = upload;
    this.source = this.resolveSource(source);
    this.cors = cors;

    // Handle injects/rejects
    this.source.inject(injections);
    this.source.reject(rejections);
    this.source.rejectDeep(rejectionsDeep);

    // Configurate integrations
    this.configureValidators(validators);
    this.configureAuthorizers(authorizers);
    this.configurePaths(integrations);

    // Finally expose the definition for CDK/CloudFormation
    this.exposeDefinition();
  }


  private configureValidators(validators: Record<string, Validator>): void {

    const defaults = Object.keys(validators).filter(k => validators[k].default === true)

    // Ensure only single validator configured as default
    if (defaults.length > 1) {
      addError(this.scope, `You may only configure one default validator`);
      return;
    }

    // Configure the default validator if provided
    if (defaults.length === 1) {
      this.source.set('x-amazon-apigateway-request-validator', defaults[0])
    }

    const cleaned = <Record<string, XAmazonApigatewayRequestValidator>>omitDeep(validators, 'default')

    this.source.set('x-amazon-apigateway-request-validators', cleaned);
  }

  /**
   * Configure Authorizers within OpenApi `components.securitySchemes`.
   */
  private configureAuthorizers(authorizers: SecuritySchemes): void {
    Object.keys(authorizers).map(id => {
      const config = authorizers[id];
      const path = `components.securitySchemes.${id}`;

      if (!this.source.has(path)) {
        addError(this.scope, `Security Scheme ${id} not found in OpenAPI Definition`);
        return;
      }

      const authorizer = this.source.get<Authorizer>(path);
      authorizer['x-amazon-apigateway-authtype'] = config['x-amazon-apigateway-authtype'];
      authorizer['x-amazon-apigateway-authorizer'] = config['x-amazon-apigateway-authorizer'];
      this.source.set(path, authorizer);
    });
  }

  /**
   * Configure all `x-amazon-apigateway-integration` values within OpenApi `paths`.
   */
  private configurePaths(paths: OpenApiPathIntegrations): void {
    Object.keys(paths).map(path => {
      if (!this.source.has(path)) {
        addError(this.scope, `OpenAPI schema is missing path for: ${path}`);
        return;
      }

      if (typeof this.cors !== 'undefined') {
        this.source.set(`${path}.OPTIONS`, this.cors);
      }

      const methods = paths[path];
      this.configurePathMethods(path, methods);
    });
  }

  /**
   * Configure `x-amazon-apigateway-integration` for given method.
   */
  private configurePathMethods(path: string, methods: OpenApiMethodIntegrations): void {
    Object.keys(methods).map(m => {
      const method = m.toLowerCase();
      this.ensureMethodExists(path, method);
      this.ensureNoIntegrationAlready(path, method);

      const integration = methods[method.toUpperCase() as Method]!;

      const methodPath = `paths['${path}']['${method}']`;

      if (typeof integration.xAmazonApiGatewayRequestValidator === 'string') {
        this.source.set(`${methodPath}['x-amazon-apigateway-request-validator']`, integration.xAmazonApiGatewayRequestValidator);
      }

      this.source.set(`${methodPath}['x-amazon-apigateway-integration']`, integration.xAmazonIntegration);
    });
  }

  /**
   * End-User can provide the OpenAPI definition either as a path to a file or
   * as an Asset. This method handles that and always returns Asset Source.
   */
  private resolveSource(source: string | Schema): Schema {
    if (typeof source === 'string') {
      return Schema.fromAsset(source);
    }
    return source;
  }


  /**
   * Set `s3Location` or `definition` to expose the generated definition
   * for CDK/CloudFormation.
   */
   private exposeDefinition(): void {
    const newSchema = this.source.definition;
    if (this.upload) {
      const newSchemaPath = __dirname + 'open-api-schema.compiled.yaml';
      fs.writeFileSync(newSchemaPath, JSON.stringify(newSchema), 'utf-8');
      const asset = new assets.Asset(this.scope, 'DefinitionAsset', {
        path: newSchemaPath,
      });

      this.s3Location = {
        bucket: asset.bucket.bucketArn,
        key: asset.s3ObjectKey,
      };
    } else {
      this.definition = newSchema;
    }
  }

  /**
   * Implement the functionality of exposing the API definition, either as S3 Asset Location or as Inline Definition.
   * @see https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-apigateway/lib/api-definition.ts#L134-L228
   */
  bind(_: Construct): apigateway.ApiDefinitionConfig {
    if (this.upload) {
      return {
        s3Location: this.s3Location,
      };
    }
    return {
      inlineDefinition: this.definition,
    };
  }

  /**
   * Ensures OpenAPI definition contains a given method for the path.
   */
  private ensureMethodExists(path: string, method: string): void {
    const value = this.source.get(`paths[${path}][${method}]`);

    if (typeof value === 'undefined') {
      addError(this.scope, `OpenAPI schema is missing method ${method} for path: ${path}`);
    }
  }

  /**
   * Ensures OpenAPI definition does not already have
   * `x-amazon-apigateway-integration` configuration for given method for the path.
   */
  private ensureNoIntegrationAlready(path: string, method: string): void {
    const value = this.source.get(`paths[${path}][${method}]['x-amazon-apigateway-integration']`);
    if (typeof value !== 'undefined') {
      addError(this.scope, `OpenAPI schema already has x-amazon-apigateway-integration configuration for method ${method} in path: ${path}`);
    }
  }

}
