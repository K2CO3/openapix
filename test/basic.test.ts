import * as cdk from 'aws-cdk-lib';
import { Duration } from 'aws-cdk-lib';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import { get } from 'lodash';
import * as openapix from '../src';

test('Basic usage', () => {
  const stack = new cdk.Stack();
  const { definition } = new openapix.OpenApi(stack, 'MyApi', {
    upload: false,
    source: new openapix.Schema({
      openapi: '3.0.1',
    }),
  });

  const actual = get(definition, 'definition');


  expect(actual).toEqual({ openapi: '3.0.1' } );
});

test('Inject paths', () => {
  const stack = new cdk.Stack();
  const { definition } = new openapix.OpenApi(stack, 'MyApi', {
    upload: false,
    source: new openapix.Schema({
      openapi: '3.0.1',
      paths: {
        '/foo': {
          GET: {
            operationId: 'get-foo',
            responses: {
              200: {
                content: {
                  'application/json': {
                    example: [
                      {
                        some: 'foo',
                        thing: 'bar',
                      },
                    ],
                  },
                },
                description: 'foo',
              },
            },
          },
        },
      },
    }),
    injections: {
      baz: 1,
    },
  });

  expect(get(definition, 'definition.baz')).toBe(1);
});

test('Reject deep paths', () => {
  const stack = new cdk.Stack();
  const { definition } = new openapix.OpenApi(stack, 'MyApi', {
    upload: false,
    source: new openapix.Schema({
      openapi: '3.0.1',
      paths: {
        '/foo': {
          GET: {
            operationId: 'get-foo',
            responses: {
              200: {
                content: {
                  'application/json': {
                    example: [
                      {
                        some: 'foo',
                        thing: 'bar',
                      },
                    ],
                  },
                },
                description: 'foo',
              },
            },
          },
        },
      },
    }),
    rejectionsDeep: ['example'],
  });


  expect(get(definition, 'definition.paths./foo.get.responses."200".content."application/json".example')).toBeUndefined();
});

test('Handles custom authorizer', () => {
  const stack = new cdk.Stack();
  const testLambda = new lambda.Function(stack, 'TestFunction', {
    runtime: lambda.Runtime.NODEJS_14_X,
    handler: 'index.handler',
    code: lambda.Code.fromInline(`module.exports = {
              handler: async (event) => {
                console.log(event);
                return {
                  statusCode: 200,
                  body: JSON.stringify({
                    message: 'Hello',
                  }),
                };
              },
            }`),
  });

  const authorizerName = 'MyLambdaAuthorizer';

  const { definition } = new openapix.OpenApi(stack, 'MyApi', {
    upload: false,
    source: new openapix.Schema({
      openapi: '3.0.1',
      components: {
        securitySchemes: {
          [authorizerName]: {
            type: 'apiKey',
            name: 'code',
            in: 'query',
          },
        },
      },
      paths: {
        '/foo': {
          GET: {
            operationId: 'get-foo',
            security: [
              {
                [authorizerName]: [],
              },
            ],
            responses: {
              200: {
                content: {
                  'application/json': {
                    example: [
                      {
                        some: 'foo',
                        thing: 'bar',
                      },
                    ],
                  },
                },
                description: 'foo',
              },
            },
          },
        },
      },
    }),

    authorizers: [

      new openapix.LambdaAuthorizer(stack, authorizerName, {
        fn: testLambda,
        identitySource: apigateway.IdentitySource.queryString('code'),
        type: 'request',
        authType: 'custom',
        resultsCacheTtl: Duration.minutes(5),
      }),
    ],

    paths: {
      '/foo': {
        GET: new openapix.LambdaIntegration(stack, testLambda),
      },
    },
  });

  const actual = get(definition, 'definition');

  console.log('foo.get', actual.paths['/foo'].GET);

  expect(get(actual, 'components.securitySchemes.MyLambdaAuthorizer')).toBeDefined();
  expect(get(actual, 'paths./foo.GET.security[0].MyLambdaAuthorizer')).toEqual([]);
});
