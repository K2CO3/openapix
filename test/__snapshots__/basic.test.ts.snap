// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Synth 1`] = `
Object {
  "Outputs": Object {
    "ExportsOutputFnGetAttTestFunction22AD90FCArn01516C4C": Object {
      "Export": Object {
        "Name": "TestLambdaStack:ExportsOutputFnGetAttTestFunction22AD90FCArn01516C4C",
      },
      "Value": Object {
        "Fn::GetAtt": Array [
          "TestFunction22AD90FC",
          "Arn",
        ],
      },
    },
  },
  "Parameters": Object {
    "BootstrapVersion": Object {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": Object {
    "TestFunction22AD90FC": Object {
      "DependsOn": Array [
        "TestFunctionServiceRole6ABD93C7",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "module.exports = {
      handler: async (event) => {
        console.log(event);
        return {
          statusCode: 200,
          body: JSON.stringify({
            message: 'Hello',
          }),
        };
      },
    }",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "TestFunctionServiceRole6ABD93C7",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "TestFunctionServiceRole6ABD93C7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
